docs/LanguageGuide.md

```diff
@@ -227,6 +156,88 @@ These can only be used at the REPL and during macro evaluation. Here's a subset
 
 To see all functions available in the `Dynamic` module, enter `(info Dynamic)` at the REPL.
 
+
+### Structs
+Any structure type defined in Carp has an init method that can be used to create a new instance. It must be called with all the arguments in the order they are defined.
+```clojure
+(deftype Vector2 [x Int, y Int])
+
+(let [my-pos (Vector2.init 10 20)]
+  ...)
+
+;; Additionally, a 'lens' is automatically generated for each member; signatures for reference:
+;; Vector2.x (Fn [(Ref Vector2)] (Ref Int))
+(Vector2.x &my-pos) ;; => 10
+;; Vector2.set-x (Fn [Vector2 Int] Vector2)
+(Vector2.set-x my-pos 30) ;; => (Vector2 30 20)
+;; Vector2.set-x! (Fn [(Ref Vector2), Int] ())
+(Vector2.set-x! &my-pos 30) ;; => Will update the vector my-pos in place and return ()
+;; Note the inner reference to a function
+;; Vector2.update-x (Fn [Vector2, (Ref (Fn [Int] Int))] Vector2)
+(Vector2.update-x my-pos inc) ;; => (Vector2 11 20)
+;; This can also be a lambda
+(Vector2.update-x my-pos &(fn [n] (* n 3))) ;; => (Vector2 30 20)
+```
+
+### Sumtypes
+There are two ways to define `sumtypes`:
+
+**Enumeration:**
+```clojure
+(deftype MyEnum
+  Kind1
+  Kind2
+  Kind3)
+```
+
+**Data:**
+```clojure
+(deftype (Either a b)
+  (Left [a])
+  (Right [b]))
+```
+
+A Variant can be created with the same syntax as call expression:
+```clojure
+(MyEnum.Kind1)
+(Either.Left 10)
+(Either.Right 11)
+
+;; Or use `use` statement
+(use Either)
+(Left 10)
+(Right 11)
+
+(use MyEnum)
+(Kind1)
+(Kind2)
+(Kind3)
+```
+
+You can use pattern matching to extract values in a safe way:
+```clojure
+(defn get [either]
+  (match either
+    (Either.Left a) a
+    (Either.Right b) b))
+
+(with MyEnum
+  ;; You can give a generic "otherwise" statement as well
+  (match myenum
+    (Kind1) (logic1)
+    _ (logic-other)))
+```
+
+Note that match works with *values* (not references) takes ownership over the value being matched on. If you instead want to match on a reference, you can use `match-ref`:
```

Comment:
I think backticks around `match` would be good here, and that there is an `and` missing after the parentheses.